### PR TITLE
[REVIEW] Remove max_strings and max_chars from nvtext::subword_tokenize

### PR DESCRIPTION
**Pull request by davidwendt**
This PR helps address some of the concerns in #5799 
The `max_num_strings` and `max_num_chars` were causing confusion an inconsistency as also mentioned in #5796 
As the current API is written, there is no need for these parameters as ported from CLX.
Using a reduced set of strings is possible by simply slicing the `column_view` of the input strings column.

The new libcudf API looks like this with 2 above parameters removed:
```
tokenizer_result subword_tokenize(
  cudf::strings_column_view const& strings,
  std::string const& filename_hashed_vocabulary,
  uint32_t max_sequence_length,
  uint32_t stride,
  bool do_lower_case,
  bool do_truncate,
  uint32_t max_rows_tensor,
  rmm::mr::device_memory_resource* mr);
```

The `max_rows_tensor` is still required since it is used along with `do_truncate` to control the desired output layout.

The corresponding Python API was also updated.

The side effect is a reduced memory footprint since only the input strings are used to allocate the working memory buffers. Also included in this PR is moving the allocation of the working memory (mostly device_uvectors) to reduce the calls to `resize()`. Finally the cub calls were replaced with thrust to simplify the logic and reduce the number of working buffers to manage. Overall these changes improved performance of the subword-tokenize gbenchmark by about ~3x though the test dataset used there is artificially simple.

### CODE REVIEW

**Comment by davidwendt** [comment]
rerun tests

**Comment by davidwendt** [comment]
Changes/fixes here will also close #5812

File: cpp/src/text/subword/detail/data_normalizer.hpp [hunk 1]
```diff
@@ -41,15 +35,20 @@ struct ptr_length_pair {
  * Converts characters to lowercase, adds spaces around punctuation and multi-byte
  * characters, strips accents from letters in the text and standardizes whitespace
  * characters to all be the code point for the " " literal.
+ *
+ * The algorithm produces tow vectors of integers `uvector_pair`.
```
Comment (vuule) [inline cpp/src/text/subword/detail/data_normalizer.hpp:39]:
```suggestion
 * The algorithm produces two vectors of integers `uvector_pair`.
```

File: cpp/src/text/subword/detail/tokenizer_utils.cuh [hunk 2]
```diff
@@ -16,39 +16,61 @@
...
-    d_lengths[idx] = d_chars_up_to_idx[d_lengths[idx] - 1];
+    auto const offset = d_offsets[idx];
+    d_offsets[idx]    = offset > 0 ? d_chars_up_to_idx[offset - 1] : 0;
   }
 };
 
+/**
+ * @brief Retrieve the code point metadata table.
+ *
+ * This is a singleton instance that copies are large table of integers into
```
Comment (vuule) [inline cpp/src/text/subword/detail/tokenizer_utils.cuh:58]:
```suggestion
 * This is a singleton instance that copies a large table of integers into
```

File: cpp/src/text/subword/detail/tokenizer_utils.cuh [hunk 3]
```diff
@@ -16,39 +16,61 @@
...
+ * device memory on the very first call.
+ *
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ */
+codepoint_metadata_type const* get_codepoint_metadata(cudaStream_t stream);
+
+/**
+ * @brief Retrieve the aux code point metadata table.
+ *
+ * This is a singleton instance that copies are large table of integers into
```
Comment (vuule) [inline cpp/src/text/subword/detail/tokenizer_utils.cuh:68]:
```suggestion
 * This is a singleton instance that copies a large table of integers into
```

File: cpp/src/text/subword/subword_tokenize.cu [hunk 4]
```diff
@@ -113,7 +110,7 @@ __global__ void kernel_compute_tensor_metadata(
       metadata[metadata_idx + 1] = (max_sequence_length - stride) / 2;
     if (last_row_of_tensor) {
       if (n_tokens_tensor < max_sequence_length)
-        metadata[metadata_idx + 2] = n_tokens_tensor - 1;
+        metadata[metadata_idx + 2] = (n_tokens_tensor > 0) ? (n_tokens_tensor - 1) : 0;
```
Comment (vuule) [inline cpp/src/text/subword/subword_tokenize.cu:113]: was there a bug when `n_tokens_tensor == 0`?

File: cpp/src/text/subword/wordpiece_tokenizer.cu [hunk 5]
```diff
@@ -276,59 +278,40 @@ wordpiece_tokenizer::wordpiece_tokenizer(hashed_vocabulary const& vocab_table,
...
+void wordpiece_tokenizer::tokenize(uvector_pair& cps_and_offsets, cudaStream_t stream)
 {
-  uint32_t* device_code_points     = cp_and_length.gpu_ptr;
-  size_t const num_code_points     = cp_and_length.length;
-  uint32_t* device_strings_offsets = offsets_and_length.gpu_ptr;
-  uint32_t const num_strings       = offsets_and_length.length - 1;
+  uint32_t* device_code_points     = cps_and_offsets.first->data();
+  size_t const num_code_points     = cps_and_offsets.first->size();
+  uint32_t* device_strings_offsets = cps_and_offsets.second->data();
+  uint32_t const num_strings       = cps_and_offsets.second->size() - 1;
```
Comment (vuule) [inline cpp/src/text/subword/wordpiece_tokenizer.cu:308]:
```suggestion
  uint32_t* const device_code_points     = cps_and_offsets.first->data();
  size_t const num_code_points           = cps_and_offsets.first->size();
  uint32_t* const device_strings_offsets = cps_and_offsets.second->data();
  uint32_t const num_strings             = cps_and_offsets.second->size() - 1;
```

File: cpp/src/text/subword/subword_tokenize.cu (hunk 4 above)
Comment (davidwendt) [reply cpp/src/text/subword/subword_tokenize.cu:113]: Yes, this was a bug. Issue #5812 mentioned in a previous comment.

File: cpp/src/text/subword/wordpiece_tokenizer.cu (hunk 5 above)
Comment (davidwendt) [reply cpp/src/text/subword/wordpiece_tokenizer.cu:308]: These 2 buffers are repurposed as working memory later in the function and so cannot be made const.

Comment (vuule) [reply cpp/src/text/subword/wordpiece_tokenizer.cu:308]: From what I can see, only the buffer content is modified, the pointers are not reseated. So the pointers can be const and point to non-const data.

Comment (davidwendt) [reply cpp/src/text/subword/wordpiece_tokenizer.cu:308]: Technically true 😊
Now I’m worried someone make the same mistake I did. I’d rather leave it as is for clarity if that is ok.

Comment (vuule) [reply cpp/src/text/subword/wordpiece_tokenizer.cu:308]: I see what you mean. The syntax is very unintuitive here. I find the const helpful here as it tells me a lot - the variable will always point to the same memory, whose content might change.
I'm not really insisting on the change, we don't really apply const that systematically. I would not have left a comment here if not for the other two consts :)

File: cpp/src/text/subword/detail/tokenizer_utils.cuh [hunk 6]
```diff
@@ -16,39 +16,61 @@
...
 
+/**
+ * @brief Retrieve the code point metadata table.
+ *
+ * This is a singleton instance that copies a large table of integers into
+ * device memory on the very first call.
+ *
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ */
+codepoint_metadata_type const* get_codepoint_metadata(cudaStream_t stream);
```
Comment (harrism) [inline cpp/src/text/subword/detail/tokenizer_utils.cuh:63]: How large? Any need to provide functions to unload and free up the memory?

File: cpp/src/text/subword/wordpiece_tokenizer.cu (hunk 5 above)
Comment (harrism) [reply cpp/src/text/subword/wordpiece_tokenizer.cu:308]: I don't think worrying that people will not know the rules is a good reason not to follow them. :)

**Review by harrism — APPROVED** [review]
I like the move to consistent Thrust use.

File: cpp/src/text/subword/detail/tokenizer_utils.cuh (hunk 6 above)
Comment (davidwendt) [reply cpp/src/text/subword/detail/tokenizer_utils.cuh:63]: The two tables are 4456KB (1114112 x 4) and 954KB (119233 x 8) so together only 5.4MB.

**Pull request merged**